Content.Shared/Burial/BurialSystem.cs

```diff
@@ -0,0 +1,168 @@
+using Content.Shared.Burial;
+using Content.Shared.Burial.Components;
+using Content.Shared.DoAfter;
+using Content.Shared.Interaction;
+using Content.Shared.Movement.Events;
+using Content.Shared.Placeable;
+using Content.Shared.Popups;
+using Content.Shared.Storage.Components;
+using Content.Shared.Storage.EntitySystems;
+using Robust.Shared.Audio.Systems;
+using Robust.Shared.Network;
+using Robust.Shared.Player;
+
+namespace Content.Server.Burial.Systems;
+
+public sealed class BurialSystem : EntitySystem
+{
+    [Dependency] private readonly SharedDoAfterSystem _doAfterSystem = default!;
+    [Dependency] private readonly SharedEntityStorageSystem _storageSystem = default!;
+    [Dependency] private readonly SharedAudioSystem _audioSystem = default!;
+    [Dependency] private readonly SharedPopupSystem _popupSystem = default!;
+    [Dependency] private readonly INetManager _netManager = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<GraveComponent, InteractUsingEvent>(OnInteractUsing);
+        SubscribeLocalEvent<GraveComponent, ActivateInWorldEvent>(OnActivate);
+        SubscribeLocalEvent<GraveComponent, AfterInteractUsingEvent>(OnAfterInteractUsing, before: new[] { typeof(PlaceableSurfaceSystem) });
+        SubscribeLocalEvent<GraveComponent, GraveDiggingDoAfterEvent>(OnGraveDigging);
+
+        SubscribeLocalEvent<GraveComponent, StorageOpenAttemptEvent>(OnOpenAttempt);
+        SubscribeLocalEvent<GraveComponent, StorageCloseAttemptEvent>(OnCloseAttempt);
+        SubscribeLocalEvent<GraveComponent, StorageAfterOpenEvent>(OnAfterOpen);
+        SubscribeLocalEvent<GraveComponent, StorageAfterCloseEvent>(OnAfterClose);
+
+        SubscribeLocalEvent<GraveComponent, ContainerRelayMovementEntityEvent>(OnRelayMovement);
+        SubscribeLocalEvent<ActiveGraveComponent, ComponentShutdown>(OnActiveGraveShutdown);
+    }
+
+    private void OnInteractUsing(EntityUid uid, GraveComponent component, InteractUsingEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        // allows someone to help you dig yourself out
+        if (TryComp<ActiveGraveComponent>(uid, out var active) && !active.DiggingSelfOut)
+            return;
+
+        if (TryComp<ShovelComponent>(args.Used, out var shovel))
+        {
+            _doAfterSystem.TryStartDoAfter(new DoAfterArgs(EntityManager, args.User, component.DigDelay / shovel.SpeedModifier, new GraveDiggingDoAfterEvent(), uid, target: args.Target, used: uid)
+            {
+                BreakOnTargetMove = true,
+                BreakOnUserMove = true,
+                BreakOnDamage = true,
+                NeedHand = true,
+                BreakOnHandChange = true
+            });
+
+            StartDigging(uid, args.User, args.Used, component);
+        }
+        else
+        {
```

Comment:
If you want exclusive do_afters you can just store the doAfterId that gets returned instead which is much simpler than tracking a component.

Reply:
Ah, that's very useful. I'll try that out!